### PR TITLE
Adding where_single_line option

### LINKED ISSUE #1936: single-line where clauses should be behind an option
`where_single_line`, false by default. And we can probably remove `where_density`.

cc https://github.com/rust-lang-nursery/fmt-rfcs/issues/74

### PR DESCRIPTION
**Pull request by afshinm**
Fixes #1936

### CODE REVIEW

File: src/items.rs [hunk 1]
```diff
@@ -2533,6 +2539,8 @@ fn rewrite_where_clause_rfc_style(
         && 6 + preds_str.len() <= shape.width
     {
         String::from(" ")
+    } else if context.config.where_single_line() {
```
Comment (nrc) [inline src/items.rs]: This looks like it could be `||`ed on to the previous `if` condition

File: src/items.rs [hunk 2]
```diff
@@ -2501,14 +2501,20 @@ fn rewrite_where_clause_rfc_style(
...
         false,
     );
-    let comma_tactic = if where_clause_option.suppress_comma {
+    let comma_tactic = if where_clause_option.suppress_comma || context.config.where_single_line() {
         SeparatorTactic::Never
     } else {
         context.config.trailing_comma()
     };
 
+    let shape_tactic = if context.config.where_single_line() {
```
Comment (nrc) [inline src/items.rs]: Do we need to check that there is only one item in the `where` clause here?

File: src/items.rs [hunk 3]
```diff
@@ -2282,11 +2282,19 @@ fn compute_budgets_for_args(
...
+        config.where_single_line(),
+    ) {
+        // where single line option
+        (_, _, true) => false,
+        (BraceStyle::AlwaysNextLine, _, _) => true,
+        (_, Density::Compressed, _) if where_clause.predicates.len() == 1 => false,
+        (_, Density::CompressedIfEmpty, _) if where_clause.predicates.len() == 1 && !has_body => {
+            false
+        }
+        (BraceStyle::SameLineWhere, _, _) if !where_clause.predicates.is_empty() => true,
```
Comment (nrc) [inline src/items.rs]: I think this code would be easier to understand if broken up into an `if` check before the match.

File: tests/target/configs-where_single_line.rs [hunk 4]
```diff
@@ -1,3 +1,12 @@
+// rustfmt-where_single_line: true
+// Where style
+
+
+fn lorem_two_items<Ipsum, Dolor, Sit, Amet>() -> T 
+where Ipsum: Eq, Lorem: Eq {
```
Comment (nrc) [inline tests/target/configs-where_single_line.rs]: This should have regular where clause formatting, not single-lined (where_single_line should only apply if there is a single item in the clause).

File: tests/target/configs-where_single_line.rs [hunk 5]
```diff
@@ -1,3 +1,12 @@
...
+    // body
+}
+
 fn lorem<Ipsum, Dolor, Sit, Amet>() -> T
```
Comment (nrc) [inline tests/target/configs-where_single_line.rs]: Could you also add a test for a single-item where clause where the rest of the signature covers multiple lines? (That should not get single-lined).

**Comment by afshinm** [comment]
Thanks for your feedbacks @nrc. I just refactored some parts of the PR and apologies for that mistake, I thought we want to inline the where clause regardless of the number of items in the where clause. 

Let me know what you think!

File: tests/target/configs-where_single_line.rs (hunk 5 above)
Comment (afshinm) [reply tests/target/configs-where_single_line.rs]: Could help me with this one? We already have one single-item where clause. What do you mean by "rest of the signature"?

Comment (nrc) [reply tests/target/configs-where_single_line.rs]: If we would put the function signature on multiple lines (e.g., if we use mutliple lines for arguments), then we should always use the multiple line form for the where clause.

File: src/items.rs [hunk 6]
```diff
@@ -2282,11 +2282,17 @@ fn compute_budgets_for_args(
...
+        return false;
+    }
     match (config.fn_brace_style(), config.where_density()) {
         (BraceStyle::AlwaysNextLine, _) => true,
-        (_, Density::Compressed) if where_clause.predicates.len() == 1 => false,
-        (_, Density::CompressedIfEmpty) if where_clause.predicates.len() == 1 && !has_body => false,
-        (BraceStyle::SameLineWhere, _) if !where_clause.predicates.is_empty() => true,
+        (_, Density::Compressed) if len == 1 => false,
+        (_, Density::CompressedIfEmpty) if len == 1 && !has_body => false,
+        (BraceStyle::SameLineWhere, _) if !is_empty => true,
```
Comment (nrc) [inline src/items.rs]: Could just use `len == 0` and avoid the `is_empty` variable.

File: src/items.rs [hunk 7]
```diff
@@ -2282,11 +2282,17 @@ fn compute_budgets_for_args(
 }
 
 fn newline_for_brace(config: &Config, where_clause: &ast::WhereClause, has_body: bool) -> bool {
+    let len = where_clause.predicates.len();
```
Comment (nrc) [inline src/items.rs]: I would call this `predicate_count` rather than `len`

File: src/items.rs [hunk 8]
```diff
@@ -2501,14 +2507,23 @@ fn rewrite_where_clause_rfc_style(
         span_end,
         false,
     );
-    let comma_tactic = if where_clause_option.suppress_comma {
-        SeparatorTactic::Never
+    let comma_tactic =
+        if where_clause_option.suppress_comma || context.config.where_single_line() && len == 1 {
```
Comment (nrc) [inline src/items.rs]: Could you factor out `context.config.where_single_line() && len == 1` into a variable?

**Comment by afshinm** [comment]
Thanks @nrc, I just refactored those parts.

> If we would put the function signature on multiple lines (e.g., if we use mutliple lines for arguments), then we should always use the multiple line form for the where clause.

So, something like:

```
fn lorem_multi_line<Ipsum, Dolor, Sit, Amet>(
    a: Aaaaaaaaaaaaaaa,
    b: Bbbbbbbbbbbbbbbb,
    c: Ccccccccccccccccc,
    d: Ddddddddddddddddddddddddd,
    e: Eeeeeeeeeeeeeeeeeee,
) -> T where Ipsum: Eq {
    // body
}
```

must be:

```
fn lorem_multi_line<Ipsum, Dolor, Sit, Amet>(
    a: Aaaaaaaaaaaaaaa,
    b: Bbbbbbbbbbbbbbbb,
    c: Ccccccccccccccccc,
    d: Ddddddddddddddddddddddddd,
    e: Eeeeeeeeeeeeeeeeeee,
) -> T 
where 
    Ipsum: Eq,
{
    // body
}
```

Is that correct?

**Comment by nrc** [comment]
> Is that correct?

Yes, correct.

**Comment by afshinm** [comment]
@nrc I'm struggling to finish the last part, to skip the where_single_line option when the rest of the signature is multi-lined.

I was thinking about doing a "\n" search and see if we have any new lines in the other parts and then assume that we have to skip (e.g. when there are multiple items in the signature). What do you think about this idea? Is there any better ways to do this?

**Comment by nrc** [comment]
> What do you think about this idea? Is there any better ways to do this?

Searching for `\n` is a fine way to do this - we do it often elsewhere.

**Comment by afshinm** [comment]
@nrc cheers Nick, let me see what can I do this weekend. :crab:

**Comment by afshinm** [comment]
@nrc OK, I just pushed the changes you've requested. Let me know what you think.

**Comment by nrc** [comment]
Awesome, thank you!

**Comment by afshinm** [comment]
Thank _you_ Nick!

File: src/config.rs [hunk 9]
```diff
@@ -574,6 +574,7 @@ create_config! {
     // function decl?
     // 2. Currently options `Tall` and `Vertical` produce the same output.
     where_density: Density, Density::Vertical, false, "Density of a where clause";
+    where_single_line: bool, false, false, "To force single line where layout";
```
Comment (remram44) [inline src/config.rs]: Both the name of the option (where single line do what?) and the "documentation" (where what is layout?) are extremely confusing. `Configurations.md` does not help; as a user I have no idea what this does.

**Pull request merged**